File: leetcode/111. Minimum Depth of Binary Tree.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        
        queue = [root]
        depth = 0
        
        while len(queue) > 0:
            depth += 1
            queue_len = len(queue)
            i = 0
            while i < queue_len:
                node = queue.pop(0)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                
                if not node.left and not node.right:
                    return depth
                
                i += 1
    
    def minDepth2(self, root):
        if not root:
            return 0
        left = self.minDepth(root.left)
        right = self.minDepth(root.right)
        return 1 + (min(left, right) or max(left, right))

File: leetcode/111. Minimum Depth of Binary Tree.js
// https://leetcode.com/problems/minimum-depth-of-binary-tree/description/

// Given a binary tree, find its minimum depth.
//
// The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
    if (!root) return 0;
    let queue = [];
    queue.push(root);
    let depth = 0;

    while (queue.length) {
        depth++;
        let queueLen = queue.length;
        let i = 0;
        while (i < queueLen) {
            let node = queue.shift();
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);

            if (!node.left && !node.right) return depth;

            i++;
        }
    }
};

var minDepth2 = function (root) {
    if (!root) return 0;
    let left = minDepth(root.left);
    let right = minDepth(root.right);
    return 1 + (Math.min(left, right) || Math.max(left, right));
};
